File: Agent/Globals/Model/PaidDeckBundleDiscount.py
import uuid


class PaidDeckBundleDiscount:
    def __init__(self, bundle_deck_id: str = None, included_deck_id: str = None, discount_percent_when_included: float = 100) -> None:
        self.__id = str(uuid.uuid4())
        self.set_bundle_deck_id(bundle_deck_id)
        self.set_included_deck_id(included_deck_id)
        self.set_discount_percent_when_included(discount_percent_when_included)

    def get_id(self) -> str:
        return self.__id

    def get_bundle_deck_id(self) -> str:
        return self.__bundle_deck_id

    def set_bundle_deck_id(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__bundle_deck_id = value

    def get_included_deck_id(self) -> str:
        return self.__included_deck_id

    def set_included_deck_id(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__included_deck_id = value

    def get_discount_percent_when_included(self) -> float:
        return self.__discount_percent_when_included

    def set_discount_percent_when_included(self, value: float) -> None:
        if value is not None:
            try:
                value = float(value)
                value = max(0, min(value, 100))
            except (ValueError, TypeError):
                value = 100
        self.__discount_percent_when_included = value

    def _restore_id_id(self, stored_id):
        if stored_id is not None:
            self.__id = stored_id

    def to_json(self) -> dict:
        return {
            'id': self.get_id(),
            'bundleDeckId': self.get_bundle_deck_id(),
            'includedDeckId': self.get_included_deck_id(),
            'discountPercentWhenIncluded': self.get_discount_percent_when_included(),
        }

    @classmethod
    def from_json(cls, data: dict) -> 'PaidDeckBundleDiscount':
        instance = cls(
            bundle_deck_id=data.get('bundleDeckId'),
            included_deck_id=data.get('includedDeckId'),
            discount_percent_when_included=data.get('discountPercentWhenIncluded')
        )
        if data.get('id') is not None:
            instance._restore_id_id(data.get('id'))
        return instance

File: Dock/Static/Globals/Model/PaidDeckBundleDiscount.js
class PaidDeckBundleDiscount
{
    #id;
    #bundleDeckId;
    #includedDeckId;
    #discountPercentWhenIncluded;

    constructor({bundleDeckId = null, includedDeckId = null, discountPercentWhenIncluded = 100} = {})
    {
        this.#id = crypto.randomUUID();
        this.setBundleDeckId(bundleDeckId);
        this.setIncludedDeckId(includedDeckId);
        this.setDiscountPercentWhenIncluded(discountPercentWhenIncluded);
    }

    getId()
    {
        return this.#id;
    }

    getBundleDeckId()
    {
        return this.#bundleDeckId;
    }

    setBundleDeckId(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#bundleDeckId = value;
    }

    getIncludedDeckId()
    {
        return this.#includedDeckId;
    }

    setIncludedDeckId(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#includedDeckId = value;
    }

    getDiscountPercentWhenIncluded()
    {
        return this.#discountPercentWhenIncluded;
    }

    setDiscountPercentWhenIncluded(value)
    {
        if (value !== null)
        {
            value = parseFloat(value);
            if (isNaN(value))
            {
                value = 100;
            }
            else
            {
                value = Math.min(Math.max(value, 0), 100);
            }
        }
        this.#discountPercentWhenIncluded = value;
    }

    toJson()
    {
        return {
            id: this.getId(),
            bundleDeckId: this.getBundleDeckId(),
            includedDeckId: this.getIncludedDeckId(),
            discountPercentWhenIncluded: this.getDiscountPercentWhenIncluded(),
        };
    }

    static fromJson(json)
    {
        const instance = new PaidDeckBundleDiscount({
            bundleDeckId: json.bundleDeckId ?? null,
            includedDeckId: json.includedDeckId ?? null,
            discountPercentWhenIncluded: json.discountPercentWhenIncluded ?? null
        });
        instance._restoreId_id(json.id);
        return instance;
    }

    _restoreId_id(storedId)
    {
        if (storedId !== undefined && storedId !== null)
        {
            this.#id = storedId;
        }
    }
}

export default PaidDeckBundleDiscount;
